strings-csharp/RandomAlphaNumericString/RandomAlphaNumericString/CryptographicallySecureMethods/SpanSecureMethod.cs

```diff
@@ -0,0 +1,36 @@
+﻿using System.Security.Cryptography;
+
+namespace RandomAlphaNumericString
+{
+    public partial class Methods
+    {
+        public static string? SpanSecureMethod(int length)
+        {
+            Span<byte> data = stackalloc byte[4 * length];
+            using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
+            {
+                rng.GetBytes(data);
+            }
+
+            Span<char> result = stackalloc char[length];
+            int charSetLength = chars.Length;
+            int acceptableByteRange = 256 - (256 % charSetLength);
+
+            int i = 0;
+            int j = 0;
+            while (i < length)
+            {
+                var b = data[j++];
+                if (b >= acceptableByteRange)
+                {
+                    b = data[j++];
+                }
+
+                result[i] = chars[b % charSetLength];
+                i++;
+            }
```

Comment:
We could make this cleaner/faster since we are specifying our character set as `a-zA-Z0-9`. We know this is 62 characters. We can mask each byte by 0x3F (63) to ensure it stays within range. We will still need to handle the case of a value over 61, but in that case we can just move to the next byte, it would be a faster calculation than the modulus.

Also, we should benchmark whether pre-generating an array of bytes is faster than just calling `Next()` and masking it for each character. I don't know which way would be faster, but it is worth the test to see and then just show the more performant one.
It may turn out that a hybrid approach is better where we use a stackalloc buffer of say Min(length, 128) and then refresh it using GetBytes() as needed.

Reply:
I tested something like this:
```
while (i < length)
{
    var b = data[j++] & 0x3F;
    if (b <= 61)
    {
        result[i++] = chars[b];
    }
}
```
This had a worse performance than using modulus. I committed a method with this code.

Also, I think I have to pre-generate the array, as there is nothing like a Next() operation.

Reply:
Hi @lennart02,
I am not sure about your benchmarking, when I ran it with the bit mask operation it was slightly faster (about 100 ns) using the bitmask vs. the method using modulus.

![image](https://github.com/CodeMazeBlog/CodeMazeGuides/assets/1489638/8ce93f53-5b86-4987-ba46-207a5f097d1c)

One thing I did change in the code was to make them match in both using RandomNumberGenerator.Fill() which is more efficient than GetBytes() when using a Span. GetBytes() will still internally create an array and then copy the data into the Span, whereas Fill() will write to the provided Span directly. That may be why you saw better performance between the modulus and the bit mask.

Regarding Next(), that was a mistake on my part, I mixed the method name between System.Random and RandomNumberGenerator, what I meant was RandomNumberGenerator.GetInt32(). Since I was running a benchmark, I ran the same using RandomNumberGenerator.GetInt32() and it was much slower than filling a stackalloc array of bytes.

Now with that being said, I looked back at the code that was using modulus and realized that we can simplify it and that leads to a method that is more performant:
![image](https://github.com/CodeMazeBlog/CodeMazeGuides/assets/1489638/6e3ab6b1-1101-4327-b129-60bd1afd65ce)

```c#
public string OldSpanSecureMethod()
{
    if (length > 128)
    {
        throw new ArgumentOutOfRangeException(nameof(length), "Length must be less than or equal to 128");
    }
    Span<byte> data = stackalloc byte[length];
    RandomNumberGenerator.Fill(data);

    Span<char> result = stackalloc char[length];
    int charSetLength = chars.Length;

    int i = 0;
    int j = 0;
    while (i < length)
    {
        result[i++] = chars[data[j++] % 62];
    }

    return new string(result);
}
```

Notice that I removed the conditional within the while loop. Since we are going to take the modulus anyway, we can use every byte that was generated. That also means that our data array can be smaller. Removing that conditional speeds things up.

Lastly, in the initial code we supported returning an arbitrary length string, now the code has been changed to only support a max of 128 characters. We could still support an arbitrary length string (and should add benchmarks for small strings (~32), medium (~128) and large (~1024)). We just need to update the code to switch from stackalloc to ArrayPool in the case of a larger string. A simple ternary check with a finally to clear and return a rented array when we have one would suffice.
You can see an example of this in the .NET source code for System.Text.Json:
https://github.com/dotnet/runtime/blob/5b5d7919a6099c9ea0056eaa800d59f53b3d9fe1/src/libraries/System.Text.Json/src/System/Text/Json/JsonEncodedText.cs#L87